R/textstat_sentiment.R

```diff
@@ -0,0 +1,136 @@
+#' assigned categories (types or features) of positive, negative, and neutral
+#' sentiment. Several formulas are available, or the user can supply a new one.
+#' @param x a [dfm] or [tokens] object containing sentiment types or features
+#' @param pos,neg character; names of the dictionary keys or features 
+#'   for positive and negative sentiment, respectively
+#' @param neut character; (optional) name of the neutral sentiment feature or 
+#'   key
+#' @param fun function; the formula for computing sentiment, which must refer to
+#'   `pos`, `neg`, and (optionally) `neut`.  Default is the Lowe et al (2001)
+#'   "logit" scale which is the log of (positive / negative) counts.  See
+#'   [sentiment-functions] for details and for additional available functions,
+#'   as well as details on how to supply custom functions.
+#' @param ... additional arguments passed to `fun`
+#' @return a data.frame of sentiment scores
+#' @export
+#' @references  Lowe, W., Benoit, K. R., Mikhaylov, S., & Laver, M. (2011).
+#'   Scaling Policy Preferences from Coded Political Texts. _Legislative Studies
+#'   Quarterly_, 36(1), 123–155.
+#'   <http://doi.org/10.1111/j.1939-9162.2010.00006.x>
+#' @examples
+#' toks <- tail(data_corpus_inaugural, n = 5) %>%
+#'     tokens(remove_punct = TRUE, remove_numbers = TRUE) %>%
+#'     tokens_lookup(dictionary = data_dictionary_LSD2015, nomatch = "_other_")
+#' textstat_sentiment(toks)
+#' textstat_sentiment(toks, smooth = 1)
+#' textstat_sentiment(toks, fun = sent_relpropdiff)
+#' 
+#' dfmat <- dfm(toks)
+#' textstat_sentiment(dfmat)
+#' 
+#' # works with multiple types associated with each sentiment category
+#' textstat_sentiment(toks, pos = c("positive", "neg_negative"),
+#'                    neg = c("negative", "neg_positive"))
+#'
+#' # with a user-supplied function
+#' sent_fn <- function(x) (x[, "pos"] - x[, "neg"]) / rowSums(x) * 100
+#' textstat_sentiment(toks, fun = sent_fn)
+textstat_sentiment <- function(x, pos = "positive", neg = "negative", 
+                               neut = NULL, fun = sent_logit, ...) {
+    UseMethod("textstat_sentiment")
+}
+
+#' @export
+textstat_sentiment.default <-  function(x, pos = "positive", neg = "negative", 
+                                        neut = NULL, fun = sent_logit, ...) {
+    stop(friendly_class_undefined_message(class(x), "textstat_sentiment"))
+}
+
+#' @export
+textstat_sentiment.tokens <- function(x, pos = "positive", neg = "negative", 
+                                      neut = NULL, fun = sent_logit, ...) {
+    textstat_sentiment(dfm(x), pos, neg, neut, fun = fun, ...)
+}
+
+#' @export
+textstat_sentiment.dfm <- function(x, pos = "positive", neg = "negative",
+                                   neut = NULL, fun = sent_logit, ...) {
+    if (!all(c(pos, neg, neut) %in% featnames(x)))
+        stop("pos, neg, and neut must match features in x")
+
+    # get the categories that are not sentiment
+    x_other <- dfm_remove(x, c(pos, neg, neut))
```

Comment:
There is `nomatch` argument in `dfm_lookup()`. I don't remember why we have it, but it might be useful here (if not we should making it useful).

Reply:
Already used it - see the comment I just added. The user has to use the lookup function before calling `textstat_sentiment()`. In this design (which is just one possible design), the user just supplies the list of types (or features)  to be associated with the positive and negative categories. These types were created from dictionary keys by `tokens_lookup()`. If the user used `nomatch = "something"` then that will be counted by a `fun` that needs an "other" count.